### PR TITLE
Use `anyOf` when `@allow-other = 'yes'` in JSON Schema

### PR DESCRIPTION
**Pull request by kylelaker**
# Committer Notes

When there is a type and an enumeration in JSON Schema, using `allOf`
means that a value must satisfy _both_ types. This is appropriate when
`@allow-other = 'no'` because it effectively limits the permitted values
to the enumeration; however, when `@allow-other = 'yes'`, that
limitation means that you are not able to use 'locally defined' values.
In this case, `anyOf` is preferable, this allows values in the
enumeration and otherwise any value that matches the schema for the
type.

A check is added explicitly for `@allow-other = 'yes'` as
`@allow-other = 'no'` is the default and matches the existing behavior.

This follows the pattern used in #253.

https://github.com/usnistgov/metaschema/blob/70085548fec75de999eebb480a0be7d7b1abdada/toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl#L328-L352

Closes: usnistgov/OSCAL#1773

### All Submissions:

- [X] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [X] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- ~Have you written new tests for your core changes, as applicable?~ @aj-stein-nist will do this in a follow-on PR
- ~Have you included examples of how to use your new feature(s)?~
- ~Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.~

### CODE REVIEW

**Comment by kylelaker** [comment]
Differences in the `OSCAL` repo after running `build/ci-cd/generate-schema.sh` with this branch can be seen at https://github.com/usnistgov/OSCAL/compare/main...kylelaker:OSCAL:oscal-jsonschema-allow-other-repro

**Comment by david-waltermire-nist** [comment]
@wendellpiez , @aj-stein-nist , and @kylelaker Do we need a new unit test to check for this?

**Comment by aj-stein-nist** [comment]
> @wendellpiez , @aj-stein-nist , and @kylelaker Do we need a new unit test to check for this?

I am looking into this more thoroughly but I am not sure what we had intended but I checked the following test (and intentionally mangled it locally to see the test generated correctly.

Given this test Metaschema module below, the following JSON schema is desired for the `allow-other="yes"` case.

https://github.com/usnistgov/metaschema/pull/253/files#diff-dae4d2728d103a48a2c7e833fd994fc629956156de0372ddd93716c16ede5a56

https://github.com/usnistgov/metaschema/pull/253/files#diff-4d22d8b5a7ea6dd9c55c7b943a106e921e9c26d491cd135d1de9e7ad00fa799b

I see how you can code around it with your PR, but this means in OSCAL something is being missed and the following code is not handling a certain condition the way it should.

https://github.com/usnistgov/metaschema/blob/70085548fec75de999eebb480a0be7d7b1abdada/toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl#L329-L338

In short, this is short-circuiting that code to avoid the bug. I guess this is an approach, but it calls into question why the test works in Metaschema but fails in the OSCAL case. I have investigated this locally on and off this evening, but I want to track down the root cause before approving this or recommending a different course of action (would also like Wendell's take on this).

Apologies for the delays, but I will continue to look into it.

**Review by aj-stein-nist — COMMENTED** [review]
@kylelaker, thanks for the PR. You can disregard my previous comments and we paired as team. It comes full circle, I now better understand why this change has to happen.

It seems we have been testing constraints on them being closed, loose, or totally open on fields, not on flags. You have fixed the part of the code we did not address in some previous work that missed this case after updating several enhancements in develop, so we really need to add some tests [to the schema generation suite](https://github.com/usnistgov/metaschema/tree/8f3d51e95c770a5e8f61a2f324e404fad412b4f1/test-suite/metaschema-xspec/json-schema-gen) to address both flags and fields.

So in summation:

1. Great work! 🎉 
2. You indicate some area of the test coverage I missed in previous work. 😢 
3. One of us needs to write tests: do you want to do it, I do it, or we work on it together? Let me know!

File: toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl [hunk 1]
```diff
@@ -194,6 +194,16 @@
             <xsl:apply-templates select="$decl/constraint/allowed-values"/>
         </xsl:variable>
```
[the remark this answers is not in the document]
Comment (wendellpiez) [reply toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:195]: This is nice! but line 196 could be

```
<xsl:variable name="allow-other" select="$enumerations/@allow-other = 'yes'"/>
```

😎 

Maybe we want to note also in a comment a link to where it says what the default is (if no value is given)?

**Comment by aj-stein-nist** [comment]
> I'm happy to read and guess at it but if you want it done well/quickly it'd probably best to either have you do it or to at least collaborate.

I am fine either way, if you want to collaborate on this, what is your week looking like? I would love to collaborate on community-authored tests (whether I drive or back-seat drive). And if that is too much for this week, totally understandable, I just PR recommended tests to your branch and we go through them?

File: toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (aj-stein-nist) [reply toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:195]: Are `'yes'` `true()` equivalent? (In other languages they call this truthiness I am not sure how to look this up in the XSLT 2.x/3.x specs.) I just wanted to be very explicit. I know your approach also works and started with that, I thought it would be best to be explicit here.

[the remark this answers is not in the document]
Comment (wendellpiez) [reply toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:195]: No, 'yes' doesn't go 'true()' as can be tested by making the cast explicit: `boolean('yes')` returns `true()` but so does `boolean('no')`. The casting rule (from XPath 1.0 in this case) is that among strings, only `''` (the empty or null string) returns `false()`.

**Comment by aj-stein-nist** [comment]
OK I guess I should get started on those tests then. :-)

File: toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (david-waltermire-nist) [reply toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:195]: Based on the feedback from @wendellpiez, how about the following?
```suggestion
        </xsl:variable>
        <xsl:variable name="allow-other" select="$enumerations/@allow-other = 'yes'"/>
        <xsl:variable name="any-or-all" select="if ($allow-other) then 'anyOf' else 'allOf'"/> 
```

File: toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl [hunk 2]
```diff
@@ -193,9 +193,11 @@
         <xsl:variable name="enumerations" as="node()*">
             <xsl:apply-templates select="$decl/constraint/allowed-values"/>
         </xsl:variable>
+        <xsl:variable name="allow-other" select="$enumerations/@allow-other = 'yes'"/>
+        <xsl:variable name="any-or-all" select="if ($allow-other) then 'anyOf' else 'allOf'"/> 
         <xsl:choose>
             <xsl:when test="exists($enumerations)">
```
Comment (david-waltermire-nist) [inline toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:199]: This should target only when the target of the allowed values is the current metaschema context. This based on the feedback from @kylelaker.

```suggestion

            <xsl:when test="$allow-other and (constraint/allowed-values/@target = '.' or empty(constraint/allowed-values/@target))">
```

**Review by david-waltermire-nist — APPROVED** [review]

**Comment by wendellpiez** [comment]
Agree with @aj-stein-nist . The changes look fine in the view and may not be likely to break anything 'worse', but the only way to prevent further thrashing is to prove out the full range of cases (with and without `@target` and with `@target='.'` in unit tests such that any of us can validate it. So let's look at expanding the XSpec coverage as well.

**Comment by aj-stein-nist** [comment]
> but the only way to prevent further thrashing is to prove out the full range of cases (with and without `@target` and with `@target='.'` in unit tests such that any of us can validate it

We do actually test for that, but we don't test the different scenarios for allowed-values for both fields and flags, that was the but I missed. I'll start work on it.

File: toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (wendellpiez) [reply toolchains/xslt-M4/schema-gen/make-json-schema-metamap.xsl:195]: It's understood with these reviews, is it not? that we also are putting those unit tests to confirm the logic is correct, not just details of syntax? I mean, I like what I'm seeing but I haven't tested it.  🤔

**Review by aj-stein-nist — CHANGES_REQUESTED** [review]
OK, so this is a fun teachable moment. @david-waltermire-nist, please revert the PR back as-is when Kyle originally submitted the PR (in [commit `6fb8a74`](https://github.com/usnistgov/metaschema/pull/360/commits/6fb8a7457ef1148589fda1de9ccc0d51d9986a97)) and I provided stylistic feedback. By doing that, I introduced a subtle bug and found it through testing #380. When testing some of the variations with the flag, I realized that we had inadvertently taken my recommendation to collapse the second if clause (or `when` rather and that is distinct but different. With #380 I now have coverage and noticed some errors I couldn't resolve until going back to the `make-json-schema-metamap.xsl` in the original.

I could find clever refactors or what not, but the real go of my previous review was to ensure I understood this submission by Kyle through tests and verify it, and in adding stuff I broke his original submission. So let's go back with that, and also add the tests with it or after. I leave that to your discretion. :-)

**Comment by aj-stein-nist** [comment]
OK, I found another edge case while touching up my tests and I cannot easily recommend the change or commit directly to Kyle's PR. I am going to recommend we close this one out, I pull in his change, and also update the relevant code in a part Kyle didn't notice to address the edge case in both places.
